feat: ensure requests can only be run one at a time

HarryET opened the pull request:
See https://linear.app/litprotocol/issue/INF-146/investigate-how-requests-are-currently-sent#comment-4976d73a for rationale.

Garandor commented on src/common/fetch.rs:
```diff
@@ -6,18 +6,26 @@ use async_std::path::PathBuf;
+
+    let _gaurd = tokio::time::timeout(Duration::from_secs(10), FETCH_LOCK.lock())
+        .await
+        .map_err(|e| error::fetch(e, Some("failed to acquire fetch lock".to_string())))?;
```
need to change all callers to actually expose this new error and distinguish timeouts from actual "nothing returned" errors so the callers can handle these differently.
Maybe introduce a new error type

HarryET replied:
added a new error type for not being able to get lock within timeout

Garandor commented on src/common/fetch.rs:
```diff
@@ -6,18 +6,26 @@ use async_std::path::PathBuf;
 ) -> error::Result<Option<Bytes>> {
     trace!("fetch_url: url: {}", url);
+
+    let _gaurd = tokio::time::timeout(Duration::from_secs(10), FETCH_LOCK.lock())
```
```suggestion
    let _guard = tokio::time::timeout(Duration::from_secs(10), FETCH_LOCK.lock())
```

Garandor replied:
you need to block the thread here, `await` is non-blocking waiting. look into `block_on`

Garandor reviewed (changes requested):
need a new error type and ensure it reaches the external caller, LGTM otherwise

Garandor commented on src/common/fetch.rs:
```diff
@@ -6,18 +6,31 @@ use async_std::path::PathBuf;
 ) -> error::Result<Option<Bytes>> {
     trace!("fetch_url: url: {}", url);
+
+    let _guard = tokio::time::timeout(Duration::from_secs(10), FETCH_LOCK.lock())
```
QQ since the mutex is wrapped in the timeout, is the _guard var actually needed or can it be 

```suggestion
    tokio::time::timeout(Duration::from_secs(10), FETCH_LOCK.lock())
```

since we're immediately awaiting on the returned future anyway

HarryET replied:
We need the gaurd so that when this function is done, the gaurd gets droped and the lock is released. Because timeout tries to resolve the value of the function in this case the lock or fail. and we need to keep the lock till the function is done

Garandor replied:
~~are you sure that's how it works? The whole promise returned from the timeout resolves in L26 due to the await. I might be wrong but I think if you want to hold a guard, to drop at the end of the fn, you need to pull out the FETCH_LOCK.lock() fn out of the timeout fn beacuse right now, the content of `_guard` is just a Promise~~
After reading the docs of timeout, i'm cool w/ this :)

Also, why are you calling it `gaurd`? once could be a typo but now i'm wondering if there's something i don't know

Garandor commented on src/common/fetch.rs:
```diff
@@ -52,11 +65,11 @@ pub async fn fetch_url(
+                            }
                         }
                     }
                 }
```
<!--__GRAPHITE_HTML_TAG_START__--><i class='graphite__hidden'>[Re: line +86]<br/><br/></i>

<!--__GRAPHITE_HTML_TAG_END__-->quick addition to help debugging here -> add `retry_sleep_ms`  to the message so we know what the AMD server returned<!--__GRAPHITE_HTML_TAG_START__--><span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/LIT-Protocol/sev-snp-utils/18?utm_source=unchanged-line-comment">Graphite</a>.</span>

<!--__GRAPHITE_HTML_TAG_END__-->

Garandor replied:
Also, this can still lock a single thread if `retry_sleep_ms` returned is large.
We should probably add a sanity check here with a fatal error log that aborts the call if we'd have to wait for more than let's say 100 minutes.

While it's not great if a thread holds this lock continuously, it's simpler than returning failure, as then new threads can send requests and we'd have to somehow prevent a cascade of new requests from extending the lockout

Please just add an warn-level log here if waiting time returned by AMD is > 1minute.

HarryET replied:
I think this was just a formatting change 🤔

HarryET replied:
Adding a warning and this to the logs sounds like a good idea though

Garandor replied:
Yeah, this wasnt a change you caused, i just noticed this going through the fn and saw this should be improved, thx for adding it :)

Garandor reviewed (changes requested):
two more quick changes and we're good to go

sidenote @HarryET, please always re-request review when you've addressed the comments so that it pops up as waiting-for-me on my Graphite

HarryET commented:
> two more quick changes and we're good to go

Just pushed my changes in response, and a response above about the gaurd.

> sidenote @HarryET, please always re-request review when you've addressed the comments so that it pops up as waiting-for-me on my Graphite

Will do!

Garandor reviewed (approved):
LGTM!
